src/messages/MDentryLink.h

```diff
@@ -17,6 +17,9 @@
 #define CEPH_MDENTRYLINK_H
 
 class MDentryLink : public Message {
+  static const int HEAD_VERSION = 1;
+  static const int COMPAT_VERSION = 1;
```

Comment:
After thinking about this more, I'm leaning towards using a common `HEAD_VERSION` and `COMPAT_VERSION` across all MDS-MDS messages (basically the "per-session version" suggested by @gregsfortytwo). I think we should either have "all MDS-MDS messages between A and B are valid" or "none are". I see potential for future problems if MDSs can only do part of a multi-message sequence (e.g. exports).

Also, having a **non-shared** per-message version means we need to handle a version for each struct we send over the wire which is unnecessarily expensive (e.g. what you're currently doing with `CDir::_encode_base`) when we're potentially sending gigabytes of replicas/exports over the wire.

Reply:
What do you think @gregsfortytwo, @jcsp, @ukernel?

Reply:
The idea sounds good to me. we just need to pass message version to various decode function, no ENCODE_START/ENCODE_END are needed.

Reply:
@batrick @ukernel  thanks for your comments, appreciated !
actually, i have thought about to using the  message's version as bufferlist version too.  the benefit is like you said, we don't need to use ENCODE_START/ENCODE_END at all.
but there are something more deep we had to consider about:
we just take CInode::_encode_base/_decode_base as an example here:
suppose we have add one filed to CInode::_encode_base. we need to do following things to make it work:
1. search how many MDS Messages have used CInode::_encode_base to encode its bufferlist.
following is what i have got through searching the code (if i have made one mistake here, something unexpected might happen, because the compiler won't tell us anything, right?): 
     a: MDSCacheRejoin
     b: MDiscoverReply
     c: MDentryLink
 2.  then change all the message's HEADER_VERSION to the same value, i think this step is not doable, each message type should have their own version, right ?
 (CInode::_encode_base is just an example, there are many other encode/decode method shared by multiple Messages)

Reply:
ENCODE_START would add 2 bytes to the bufferlist, take an example of CInode::_encode_base here too. 
  encode(first, bl);
  encode(inode, bl, features);
  encode(symlink, bl);
  encode(dirfragtree, bl);
  encode(xattrs, bl);
  encode(old_inodes, bl, features);
  encode(damage_flags, bl);
  encode_snap(bl);
compared to the data of CInode::_encode_base here, is 2 bytes version info really matters ?
 
I mean if it's a feature we must have, then it's the little cost we must take.

Reply:
> search how many MDS Messages have used CInode::_encode_base to encode its bufferlist.

No, we can just inrecement a `#define CEPH_MDS_MESSAGE_VERSION` in a common header. That define is used by all the MDS-to-MDS messages'  `HEAD_VERSION`.

> compared to the data of CInode::_encode_base here, is 2 bytes version info really matters ?

If we're sending 100k+ inodes during an export, it would be begin to have an effect. At the very least, it adds unnecessary branches for each version test. Ultimately though, you need to ask yourself what is the benefit of allowing some MDS-to-MDS messages to be processed vs. allowing others. (I don't think there is a benefit... in fact it could result in difficult to diagnose problems.)